---
title: "Explore_intersect_ac_m3_boolean"
author: "KN"
date: "1/25/2023"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown
```{r setup2}

library(dplyr)
library(ggplot2)
library(ggpubr)
library(gridExtra)
library(data.table)
#install.packages("VennDiagram")
library(VennDiagram)
library(viridis)
library(rcartocolor)
library(nord)
library("ggvenn")
```

```{r intersect_ac_m3_boolean}

peaks <- read.table("H3K27ac.consensus_peaks.boolean.annotatePeaks.txt", header=T, sep = "\t")


peaks <- peaks %>%
    mutate(Feature=gsub( " .*$", "", Annotation ))


peaks$Annotation_new <- gsub(pattern = "\\)", "", x=peaks$Annotation)
peaks$Annotation_new <- gsub(pattern = "\\(", "", x=peaks$Annotation_new)

annot.df <- as.data.frame(do.call("rbind", strsplit(peaks$Annotation_new, " ")))[,c(1,2,4,6)]
colnames(annot.df) <- c("feat", "gene", "pos_number", "tot_exon_intr" )

peaks <- cbind(peaks, annot.df)

#for intersect with m3
peaks_ac <- 
  peaks %>%
  filter(num_samples == 4)

peaks_2kb_ac <- 
peaks_ac %>% 
  filter(Distance.to.TSS < 2000) %>%
  filter(Distance.to.TSS > -2000)


#get m3 data
peaks <- read.table("H3K4m3.consensus_peaks.boolean.annotatePeaks.txt", header=T, sep = "\t")


peaks <- peaks %>%
    mutate(Feature=gsub( " .*$", "", Annotation ))


peaks$Annotation_new <- gsub(pattern = "\\)", "", x=peaks$Annotation)
peaks$Annotation_new <- gsub(pattern = "\\(", "", x=peaks$Annotation_new)

annot.df <- as.data.frame(do.call("rbind", strsplit(peaks$Annotation_new, " ")))[,c(1,2,4,6)]
colnames(annot.df) <- c("feat", "gene", "pos_number", "tot_exon_intr" )

peaks <- cbind(peaks, annot.df)

#for intersect with ac
peaks_m3 <- 
  peaks %>%
  filter(num_samples == 4)

peaks_2kb_m3 <- 
peaks_m3 %>% 
  filter(Distance.to.TSS < 2000) %>%
  filter(Distance.to.TSS > -2000)


#intersect
x = data.table(peaks_ac[,1:4])
y = data.table(peaks_m3[,1:3])
setkey(y, chr, start, end)

foverlaps(x, y, type = "any")
foverlaps(x, y, type = "any", nomatch = NULL)
#5383

#for venn diagram

H3K27ac=length(peaks_ac$chr)
H3K4me3=length(peaks_m3$chr)
inter_ac_m3=length(foverlaps(x, y, type = "any", nomatch = NULL)[,1]$chr)

y_2kb = data.table(peaks_2kb_m3[,1:3])
setkey(y_2kb, chr, start, end)
inter_ac_m3_2kb=length(foverlaps(x=data.table(peaks_2kb_ac[,1:4]), y=y_2kb, type = "any", nomatch = NULL)[,1]$chr)

venn_diag <- 
draw.pairwise.venn(area1=H3K27ac, area2=H3K4me3, cross.area=inter_ac_m3, 
                   category = c(paste("H3K27ac \n(n=", length(peaks_ac$chr), ")", sep = ""), paste("H3K4me3\n(n=", length(peaks_m3$chr), ") ", sep = "")), 
                   lty = rep("blank", 2),
                   fill = c("#526A83", "#526A83"),
                   alpha = c(0.3, 0.6),
                   cex = 1.8,
                   cat.pos = c(0,0),     #text dist from centre
                   cat.dist = c(0.04,0.072),     #text dist from rim of circle
                   cat.cex = 1.8,
                   fontfamily = rep("sans", 3),
                   cat.fontfamily = c("sans", "sans"))

#colours
venn_diag <- 
draw.pairwise.venn(area1=H3K27ac, area2=H3K4me3, cross.area=inter_ac_m3, 
                   category = c(paste("H3K27ac \n(n=", length(peaks_ac$chr), ")", sep = ""), paste("H3K4me3\n(n=", length(peaks_m3$chr), ") ", sep = "")), 
                   lty = rep("blank", 2),
                   fill = c(col_scheme_chip[6], col_scheme_chip[10]),
                   alpha = c(0.4, 0.7),
                   cex = 1.8,
                   cat.pos = c(0,0),     #text dist from centre
                   cat.dist = c(0.04,0.072),     #text dist from rim of circle
                   cat.cex = 1.8,
                   fontfamily = rep("sans", 3),
                   cat.fontfamily = c("sans", "sans"))


venn_diag <- 
draw.pairwise.venn(area1=H3K27ac, area2=H3K4me3, cross.area=inter_ac_m3, 
                   category = c(paste("H3K27ac \n(n=", length(peaks_ac$chr), ")", sep = ""), paste("H3K4me3\n(n=", length(peaks_m3$chr), ") ", sep = "")), 
                   lty = rep("blank", 2),
                   fill = c(col_scheme_chip[6], col_scheme_chip[10]),
                   alpha = c(0.4, 0.7),
                   cex = 1.8,
                   cat.pos = c(0,0),     #text dist from centre
                   cat.dist = c(0.07,-0.4),     #text dist from rim of circle
                   cat.cex = 1.8,
                   fontfamily = rep("sans", 3),
                   cat.fontfamily = c("sans", "sans"),
                   rotation.degree = -90,
                   margin = 0.06)

#colours
venn_diag_2kb <- 
draw.pairwise.venn(area1=length(peaks_2kb_ac$chr), area2=length(peaks_2kb_m3$chr), cross.area=inter_ac_m3_2kb, 
                   category = c(paste("H3K27ac (< 2Kb) \n(n=", length(peaks_2kb_ac$chr), ")", sep = ""), paste("H3K4me3 (< 2Kb)\n(n=", length(peaks_2kb_m3$chr), ") ", sep = "")), 
                   lty = rep("blank", 2),
                   fill = c(col_scheme_chip[6], col_scheme_chip[10]),
                   alpha = c(0.4, 0.7),
                   cex = 1.8,
                   cat.pos = c(0,0),     #text dist from centre
                   cat.dist = c(-0.62,0.24),     #text dist from rim of circle
                   cat.cex = 1.8,
                   fontfamily = rep("sans", 3),
                   cat.fontfamily = c("sans", "sans"),
                   rotation.degree = 90,
                   margin = 0.095)


tiff(
  filename = "venn_diag.tiff",
  compression = "lzw",
  height = 6,
  width = 6);
grid.draw(venn_diag);
grid.newpage()
  dev.off();
  
ggsave(
  ggarrange(venn_diag, venn_diag_2kb),
  filename = "venn_diag_2.pdf",
  device = "pdf",
  height = 8,
  width = 16)


scales::show_col(c( "#A0885C", "#70495c", "#526A83", "#BD434E", "#F3A250", "#49705D","#A4B7AE", "#b7a4ad", "#b8a5a5", "#8a6262", "#e0b4b4"))
  
col_scheme_chip <- c( "#A0885C", "#70495c", "#526A83", "#BD434E", "#F3A250", "#49705D","#A4B7AE", "#b7a4ad", "#b8a5a5", "#8a6262", "#e0b4b4")

#dark green
col_scheme_chip[6]

#dark pink
col_scheme_chip[10]

```

```{r total_nr_exons}

nr_exons <- read.table("numberOfExonsTotal.txt", header = F)

nr_exons$V3 <- colSums(nr_exons)[1] - cumsum(lag(nr_exons$V1, default = 0))

nr_exons %>%
ggplot() +
  geom_bar(aes(x = V2, y = V3), stat="identity")


#counts per exon position
peaks_per_exon_nr_ac <- 
peaks_ac %>%
  filter(feat == "exon") %>%
  count(pos_number) 


peaks_per_exon_nr_m3 <- 
peaks_m3 %>%
  filter(feat == "exon") %>%
  count(pos_number) 

nr_exons$V2 <- as.character(nr_exons$V2)

nr_exon_ac_m3 <- left_join(left_join(nr_exons, peaks_per_exon_nr_ac, by = c("V2" = "pos_number")), peaks_per_exon_nr_m3, by = c("V2" = "pos_number"))


nr_exon_ac_m3 %>%
ggplot() +
  geom_bar(aes(x = as.numeric(V2), y = n.x/V3), stat="identity")


nr_exon_ac_m3 %>%
ggplot() +
  geom_bar(aes(x = as.numeric(V2), y = n.y/V3), stat="identity")

nr_exon_ac_m3$feat <- "exon"

#intron
#counts per exon position
peaks_per_intron_nr_ac <- 
peaks_ac %>%
  filter(feat == "intron") %>%
  count(pos_number) 


peaks_per_intron_nr_m3 <- 
peaks_m3 %>%
  filter(feat == "intron") %>%
  count(pos_number) 

nr_exons$V2 <- as.character(nr_exons$V2)

nr_introns_ac_m3 <- left_join(left_join(nr_exons, peaks_per_intron_nr_ac, by = c("V2" = "pos_number")), peaks_per_intron_nr_m3, by = c("V2" = "pos_number"))


nr_introns_ac_m3$feat <- "intron"

nr_ac_m3 <- rbind(nr_exon_ac_m3, nr_introns_ac_m3)

nr_ac_plot <- 
nr_ac_m3 %>%
  ggplot(aes(reorder(nr_ac_m3$V2, as.numeric(nr_ac_m3$V2)), 100*n.x/V3, fill=feat)) +
  geom_bar(stat = "identity", position="dodge", alpha=0.75) +
  scale_fill_manual(values = c("#A0885C", "#526A83"),
                    name="",
                   labels=c("Exon", "Intron")) +
  scale_y_continuous(limits = c(0,10)) +
  scale_x_discrete(limits = c(seq(1,20))) +
  labs(title=paste("H3K27ac (n=", length(peaks_ac$chr), ")", sep = ""),
       x="Exon/intron position",
       y="Percent exon/intron with peak (%)"
       #tag = "c"
       ) +
  theme_pubr() +
  theme(plot.title = element_text(size = 12, colour = "black", hjust = 0.55, vjust = -5),
        axis.text = element_text(size = 12, colour = "black"),
        axis.title = element_text(size = 12, colour = "black"))

nr_m3_plot <- 
nr_ac_m3 %>%
  ggplot(aes(reorder(nr_ac_m3$V2, as.numeric(nr_ac_m3$V2)), 100*n.y/V3, fill=feat)) +
  geom_bar(stat = "identity", position="dodge", alpha=0.75) +
  scale_fill_manual(values = c("#A0885C", "#526A83"),
                    name="",
                   labels=c("Exon", "Intron"), 
                   guide = "none") +
  scale_y_continuous(limits = c(0,10)) +
  scale_x_discrete(limits = c(seq(1,20))) +
  labs(title=paste("H3K4me3 (n=", length(peaks_m3$chr), ")", sep = ""),
       x="Exon/intron position",
       y="Percent exon/intron with peak (%)"
       #tag = "d"
       ) +
  theme_pubr() +
  theme(plot.title = element_text(size = 12, colour = "black", hjust = 0.55, vjust = -5),
        axis.text = element_text(size = 12, colour = "black"),
        axis.title = element_text(size = 12, colour = "black"))

feat_labels <- c("Exon", "Intergenic\n", "Intron", "Promoter-TSS\n", "TES")

pie_m3_ms <- 
peaks_m3 %>% 
  filter(!is.na(feat)) %>%
  group_by(feat) %>%
  count() %>%
ggplot(aes(x="", y=n, fill=feat)) +
  geom_bar(stat="identity", width=1, alpha=0.75) +
  coord_polar("y", start = 45) +
  geom_text(aes(label=paste(feat_labels, " (", n, ")", sep = ""), x=1.2), position = position_stack(vjust = 0.5), size=12/2.8) +
  theme_void() +
  theme(legend.text = element_text(size = 12),
        legend.title = element_blank()) +
  scale_fill_manual(name="Feature", 
                    values=c( "#A0885C", "#70495c", "#526A83", "#BD434E", "#F3A250"),
                    labels=c("Exon", "Intergenic", "Intron", "Promoter_TSS", "TES"))

pie_ac_ms <- 
peaks_ac %>% 
  filter(!is.na(feat)) %>%
  group_by(feat) %>%
  count() %>%
ggplot(aes(x="", y=n, fill=feat)) +
  geom_bar(stat="identity", width=1, alpha=0.75) +
  coord_polar("y", start = 45) +
  geom_text(aes(label=paste(feat_labels, " (", n, ")", sep = ""), x=1.2), position = position_stack(vjust = 0.5), size=12/2.8) +
  theme_void() +
  theme(legend.text = element_text(size = 12),
        legend.title = element_blank()) +
  scale_fill_manual(name="Feature", 
                    values=c( "#A0885C", "#70495c", "#526A83", "#BD434E", "#F3A250"),
                    labels=c("Exon", "Intergenic", "Intron", "Promoter_TSS", "TES"))

ggsave(
  ggarrange(ggarrange(nr_ac_plot + theme(legend.position = "none") + annotation_custom(ggplotGrob(pie_ac_ms + theme(legend.position = "none", plot.margin = margin(-1,-1,-1,-1, "cm"))), xmin = 2, xmax = 20, ymin = 2, ymax = 10)),
  ggarrange(nr_m3_plot + theme(legend.position = "none") + annotation_custom(ggplotGrob(pie_m3_ms + theme(legend.position = "none", plot.margin = margin(-1,-1,-1,-1, "cm"))), xmin = 2, xmax = 20, ymin = 2, ymax = 10)),
  labels = c("c", "d"), font.label = list(face = "plain"),
  nrow = 1),
       filename = "peak_distr.pdf",
       height = 6,
       width = 14)

ggsave(
  ggarrange(ggarrange(nr_ac_plot + theme(legend.position = "none") + annotation_custom(ggplotGrob(pie_ac_ms + theme(legend.position = "none", plot.margin = margin(-1,-1,-1,-1, "cm"))), xmin = 2, xmax = 20, ymin = 2, ymax = 10)),
  ggarrange(nr_m3_plot + theme(legend.position = "none") + annotation_custom(ggplotGrob(pie_m3_ms + theme(legend.position = "none", plot.margin = margin(-1,-1,-1,-1, "cm"))), xmin = 2, xmax = 20, ymin = 2, ymax = 10)),
  labels = c("c", "d"), font.label = list(face = "plain"),
  nrow = 1),
       filename = "peak_distr.png",
       height = 6,
       width = 14)

```

